Add some timing info to rustdoc

jyn514 opened the pull request:
There are various improvements, but the main one is to time each pass
that rustdoc performs (`rustdoc::passes`).

Before, these were the top five timings for `cargo doc` on the cargo
repository:

```
+---------------------------------+-----------+-----------------+----------+------------+
| Item                            | Self time | % of total time | Time     | Item count |
+---------------------------------+-----------+-----------------+----------+------------+
| <unknown>                       | 854.70ms  | 20.888          | 2.47s    | 744823     |
+---------------------------------+-----------+-----------------+----------+------------+
| expand_crate                    | 795.29ms  | 19.436          | 848.00ms | 1          |
+---------------------------------+-----------+-----------------+----------+------------+
| metadata_decode_entry           | 256.73ms  | 6.274           | 279.49ms | 518344     |
+---------------------------------+-----------+-----------------+----------+------------+
| resolve_crate                   | 240.56ms  | 5.879           | 242.86ms | 1          |
+---------------------------------+-----------+-----------------+----------+------------+
| hir_lowering                    | 146.79ms  | 3.587           | 146.79ms | 1          |
+---------------------------------+-----------+-----------------+----------+------------+
```

Now the timings are:

```
+---------------------------------+-----------+-----------------+----------+------------+
| Item                            | Self time | % of total time | Time     | Item count |
+---------------------------------+-----------+-----------------+----------+------------+
| <unknown>                       | 1.40s     | 22.662          | 3.73s    | 771430     |
+---------------------------------+-----------+-----------------+----------+------------+
| collect-trait-impls             | 1.34s     | 21.672          | 2.87s    | 1          |
+---------------------------------+-----------+-----------------+----------+------------+
| expand_crate                    | 1.21s     | 19.577          | 1.28s    | 1          |
+---------------------------------+-----------+-----------------+----------+------------+
| build extern trait impl         | 704.66ms  | 11.427          | 1.07s    | 21893      |
+---------------------------------+-----------+-----------------+----------+------------+
| metadata_decode_entry           | 354.84ms  | 5.754           | 391.81ms | 544919     |
+---------------------------------+-----------+-----------------+----------+------------+
```

The goal is to help me debug regressions like https://github.com/rust-lang/rust/pull/74518#issuecomment-661498214 (currently I have _no_ idea what could have gone wrong).

r? @eddyb or @Mark-Simulacrum

jyn514 committed:
Add some timing info to rustdoc

There are various improvements, but the main one is to time each pass
that rustdoc performs (`rustdoc::passes`).

Before, these were the top five timings for `cargo doc` on the cargo
repository:

```
+---------------------------------+-----------+-----------------+----------+------------+
| Item                            | Self time | % of total time | Time     | Item count |
+---------------------------------+-----------+-----------------+----------+------------+
| <unknown>                       | 854.70ms  | 20.888          | 2.47s    | 744823     |
+---------------------------------+-----------+-----------------+----------+------------+
| expand_crate                    | 795.29ms  | 19.436          | 848.00ms | 1          |
+---------------------------------+-----------+-----------------+----------+------------+
| metadata_decode_entry           | 256.73ms  | 6.274           | 279.49ms | 518344     |
+---------------------------------+-----------+-----------------+----------+------------+
| resolve_crate                   | 240.56ms  | 5.879           | 242.86ms | 1          |
+---------------------------------+-----------+-----------------+----------+------------+
| hir_lowering                    | 146.79ms  | 3.587           | 146.79ms | 1          |
+---------------------------------+-----------+-----------------+----------+------------+
```

Now the timings are:

```
+---------------------------------+-----------+-----------------+----------+------------+
| Item                            | Self time | % of total time | Time     | Item count |
+---------------------------------+-----------+-----------------+----------+------------+
| <unknown>                       | 1.40s     | 22.662          | 3.73s    | 771430     |
+---------------------------------+-----------+-----------------+----------+------------+
| collect-trait-impls             | 1.34s     | 21.672          | 2.87s    | 1          |
+---------------------------------+-----------+-----------------+----------+------------+
| expand_crate                    | 1.21s     | 19.577          | 1.28s    | 1          |
+---------------------------------+-----------+-----------------+----------+------------+
| build extern trait impl         | 704.66ms  | 11.427          | 1.07s    | 21893      |
+---------------------------------+-----------+-----------------+----------+------------+
| metadata_decode_entry           | 354.84ms  | 5.754           | 391.81ms | 544919     |
+---------------------------------+-----------+-----------------+----------+------------+
```

Mark-Simulacrum commented:
Do we have some idea where the unknown is coming from? The time spent in it seems to have ... gone up?

jyn514 commented:
@Mark-Simulacrum the overall time went up as well, I think because of one of the regressions in https://github.com/rust-lang/rustc-perf/blob/master/triage/2020-07-21.md. I would look at the percentages instead.

jyn514 commented:
I'm not sure where the <unknown> is coming from - I tried to wrap all of `rustdoc::core` in a timer but that didn't work because the main function doesn't have access to a compile session.

bors commented:
:umbrella: The latest upstream changes (presumably #75008) made this pull request unmergeable. Please resolve the merge conflicts.

eddyb commented:
@jyn514 You don't need to wrap the entirety of rustdoc, just the outermost thing that *has* access to a timer/profiler.

r? @Mark-Simulacrum or @wesleywiser

eddyb commented on src/librustdoc/core.rs:
```diff
@@ -402,7 +402,9 @@ pub fn run_core(options: RustdocOptions) -> (clean::Crate, RenderInfo, RenderOpt
                 let body = hir.body(hir.body_owned_by(hir.as_local_hir_id(def_id)));
                 debug!("visiting body for {:?}", def_id);
-                EmitIgnoredResolutionErrors::new(tcx).visit_body(body);
+                tcx.sess.time("emit ignored resolution errors", || {
+                    EmitIgnoredResolutionErrors::new(tcx).visit_body(body);
+                });
```
Note: not a bad change, but would not have been marked as `<unknown>` before (it would show up as part of `typeck`).

eddyb commented on src/librustdoc/core.rs:
```diff
@@ -425,16 +427,18 @@ pub fn run_core(options: RustdocOptions) -> (clean::Crate, RenderInfo, RenderOpt
-                            .unwrap_or_else(|()| {
-                                panic!("Unable to resolve external crate {}", extern_name)
-                            });
+                        sess.time("load extern crate", || {
```
I would put this around the whole loop and call it "load extern crate**s**".

eddyb commented on src/librustdoc/core.rs:
```diff
@@ -455,13 +459,17 @@ pub fn run_core(options: RustdocOptions) -> (clean::Crate, RenderInfo, RenderOpt
                 });
 
-                let access_levels = tcx.privacy_access_levels(LOCAL_CRATE);
+                let access_levels = tcx
+                    .sess
+                    .time("privacy_access_levels", || tcx.privacy_access_levels(LOCAL_CRATE));
```
This is a query, I think, so the extra `time` seems unnecessary.

eddyb commented:
At the very least, it looks like the entirety of `global_ctxt.enter(|tcx| {` could be wrapped in a named region (if not more).

jyn514 committed:
Time loading all crates, not each individual crate

jyn514 committed:
Remove redundant timing for query

jyn514 committed:
Separate `run_global_ctxt` into a separate function

jyn514 committed:
Time running the global context

jyn514 commented:
Addressed review comments. I separated the enormous closure into a separate function, so best to view with whitespace changes disabled. Building rustdoc now to see if we got rid of most of the `<unknown>`.

jyn514 committed:
xpy fmt

jyn514 commented:
Timing the global setup didn't seem to help much.
```
+-------------------------------------------------+-----------+-----------------+----------+------------+
| Item                                            | Self time | % of total time | Time     | Item count |
+-------------------------------------------------+-----------+-----------------+----------+------------+
| collect-trait-impls                             | 1.46s     | 23.619          | 3.26s    | 1          |
+-------------------------------------------------+-----------+-----------------+----------+------------+
| expand_crate                                    | 1.20s     | 19.538          | 1.25s    | 1          |
+-------------------------------------------------+-----------+-----------------+----------+------------+
| <unknown>                                       | 1.09s     | 17.746          | 3.17s    | 753829     |
+-------------------------------------------------+-----------+-----------------+----------+------------+
| run_global_ctxt                                 | 3.34ms    | 0.054           | 4.24s    | 1          |
+-------------------------------------------------+-----------+-----------------+----------+------------+

jyn514 commented:
I did find this comment though: https://github.com/rust-lang/rust/blob/62cf767a4a39b47677d18110359d9e7152dc9d1c/src/librustc_codegen_ssa/base.rs#L782-L792

jyn514 commented:
Aha, I think I found the issue: all of the html rendering isn't timed. https://github.com/rust-lang/rust/blob/1d69e3b1d753951bc7df0f02d6fd4719065d98c3/src/librustdoc/lib.rs#L525-L533

jyn514 commented:
Well, that helped a little.
```
summarize summarize time-html/unknown-crate-3300.events 
+-------------------------------------------------+-----------+-----------------+----------+------------+
| Item                                            | Self time | % of total time | Time     | Item count |
+-------------------------------------------------+-----------+-----------------+----------+------------+
| collect-trait-impls                             | 1.46s     | 21.568          | 3.29s    | 1          |
+-------------------------------------------------+-----------+-----------------+----------+------------+
| expand_crate                                    | 1.22s     | 18.103          | 1.27s    | 1          |
+-------------------------------------------------+-----------+-----------------+----------+------------+
| <unknown>                                       | 1.12s     | 16.504          | 3.35s    | 753829     |
+-------------------------------------------------+-----------+-----------------+----------+------------+
| build extern trait impl                         | 822.75ms  | 12.162          | 1.34s    | 21968      |
+-------------------------------------------------+-----------+-----------------+----------+------------+
| render html                                     | 439.39ms  | 6.495           | 439.39ms | 1          |
+-------------------------------------------------+-----------+-----------------+----------+------------+
```

jyn514 committed:
Time how long it takes to render HTML

jyn514 commented:
It's coming from `item_types_checking`. 
![flamegraph](https://user-images.githubusercontent.com/23638587/90345998-70c18880-dff3-11ea-80df-649b73121100.png)

jyn514 commented:
Anyone know why this might be getting the timing name reset?
```rust
    tcx.sess.time("item_types_checking", || {
        for &module in tcx.hir().krate().modules.keys() {
            tcx.sess.time("check_mod_item_types", || {
            tcx.ensure().check_mod_item_types(tcx.hir().local_def_id(module));
            });
        }
    });
```
`check_mod_item_types` works fine and shows almost exactly 0 time spent in it, instead there's half a second of `<unknown>`.
https://github.com/rust-lang/rust/pull/74590/files?diff=unified&w=1#diff-0de644786a2f1fd88c7d7f44bc3fa4bbR497

Mark-Simulacrum commented on src/librustdoc/core.rs:
```diff
@@ -424,18 +428,20 @@ pub fn run_core(options: RustdocOptions) -> (clean::Crate, RenderInfo, RenderOpt
-                                panic!("Unable to resolve external crate {}", extern_name)
-                            });
-                    }
+                    sess.time("load extern crates", || {
```
These should be `load_extern_crates`, we want to avoid spaces in these names for various reasons.

jyn514 replied:
Done. Note that the rustdoc passes have dashes and not underscores, is that fine?

Mark-Simulacrum replied:
I would prefer to be consistent, but dashes don't seem bad. The main reason for avoiding spaces is so that e.g. URLs don't need escaping.

Mark-Simulacrum commented:
I would personally like us to track down the problem with unknown timing before merging this in, though I don't have great ideas for doing that. Maybe @wesleywiser has suggestions?

jyn514 committed:
Use underscores instead of spaces

jyn514 commented:
Tried the following without success.
```diff
diff --git a/src/librustc_typeck/check/mod.rs b/src/librustc_typeck/check/mod.rs
index 824e81a974c..a45836f3f59 100644
--- a/src/librustc_typeck/check/mod.rs
+++ b/src/librustc_typeck/check/mod.rs
@@ -722,7 +722,9 @@ struct CheckItemTypesVisitor<'tcx> {
 
 impl ItemLikeVisitor<'tcx> for CheckItemTypesVisitor<'tcx> {
     fn visit_item(&mut self, i: &'tcx hir::Item<'tcx>) {
-        check_item_type(self.tcx, i);
+        self.tcx.sess.time("check_item_type", || {
+            check_item_type(self.tcx, i);
+        })
     }
     fn visit_trait_item(&mut self, _: &'tcx hir::TraitItem<'tcx>) {}
     fn visit_impl_item(&mut self, _: &'tcx hir::ImplItem<'tcx>) {}
```

bors commented:
:umbrella: The latest upstream changes (presumably #74098) made this pull request unmergeable. Please resolve the merge conflicts.

bors commented:
:umbrella: The latest upstream changes (presumably #74489) made this pull request unmergeable. Please resolve the merge conflicts.

jyn514 commented:
@Mark-Simulacrum I don't expect to have time to debug this soon ... are you okay with merging the current changes and opening an issue for the `<unknown>`? I'd rather have _some_ timing than having to guess.

jyn514 committed:
Fix bad rebase

Had to use -Xignore-whitespace-conflicts to avoid awful conflicts, but
that threw off the indentation

wesleywiser commented:
I've only skimmed the PR but I think we should not block merging this with resolving the `unknown` issue. We've seen that before in other profiles so it's probably another instance of the existing issue. I hope to have some time this week to look into it.

Mark-Simulacrum commented:
Okay, knowing that we've seen this before is indeed helpful and that convinces me that it should be fine to move ahead here.

@bors r+

bors commented:
:pushpin: Commit 6e3e74dd719334e84edd70ba0a842f5d4f94d266 has been approved by `Mark-Simulacrum`

<!-- @bors r=Mark-Simulacrum 6e3e74dd719334e84edd70ba0a842f5d4f94d266 -->
<!-- homu: {"type":"Approved","sha":"6e3e74dd719334e84edd70ba0a842f5d4f94d266","approver":"Mark-Simulacrum"} -->

bors commented:
:hourglass: Testing commit 6e3e74dd719334e84edd70ba0a842f5d4f94d266 with merge 03017003c77d782cf7ed841a05d7c628a9b93f25...
<!-- homu: {"type":"BuildStarted","head_sha":"6e3e74dd719334e84edd70ba0a842f5d4f94d266","merge_sha":"03017003c77d782cf7ed841a05d7c628a9b93f25"} -->

camelid commented:
You may already know this, but this seems to be the rustc code that produces the `<unknown>` name:

https://github.com/rust-lang/rust/blob/192c7dbb6dbd0b2c176101ed2fe785901b2a457d/compiler/rustc_codegen_llvm/src/debuginfo/metadata.rs#L795

Not sure why it's happening though.

jyn514 commented:
That's under debuginfo, so I wouldn't expect it to be related.

I did notice a while back that rustdoc doesn't include the crate name in the .events file - I think that might be because doctree passes None for the crate name? I need to test it.
https://github.com/rust-lang/rust/blob/master/src/librustdoc/visit_ast.rs#L71

jyn514 commented:
> I did notice a while back that rustdoc doesn't include the crate name in the .events file - I think that might be because doctree passes None for the crate name? I need to test it.

This was https://github.com/rust-lang/rust/pull/79586.
